code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -305,24 +305,27 @@
 		activate_sr()
 
 
-/obj/item/borg/upgrade/selfrepair/update_icon()
+/obj/item/borg/upgrade/selfrepair/update_icon_state()
 	if(toggle_action)
 		icon_state = "selfrepair_[on ? "on" : "off"]"
-		for(var/X in actions)
-			var/datum/action/A = X
-			A.UpdateButtonIcon()
 	else
 		icon_state = "cyborg_upgrade5"
 
 /obj/item/borg/upgrade/selfrepair/proc/activate_sr()
 	START_PROCESSING(SSobj, src)
 	on = TRUE
 	update_icon()
+	for(var/X in actions)
+		var/datum/action/A = X
+		A.UpdateButtonIcon()

```

Comment:
You made a signal for this sort of thing

Reply:
I'm gonna have to do this carefully but sure, I'll get around to it

Reply:
Actually I'm not sure what to do here, have the action button hook into the updated icon signal to call UpdateButtonIcon()?

Reply:
yeah exactly